libs/cardano-ledger-test/bench/Bench/Cardano/Ledger/StakeDistr.hs

```diff
@@ -0,0 +1,388 @@
+import Data.Strings (lazyBytes)
+import Data.Text (pack)
+import qualified Data.UMap as UM
+import GHC.IO.Exception (IOException (..))
+import Test.Cardano.Ledger.Tickf (oldCalculatePoolDistr)
+
+-- ======================================================================
+
+type CurrentEra = AlonzoEra StandardCrypto
+
+-- ==============================
+-- Data files
+
+-- | A filepath to a CBOR file that can be decoded to a NewEpochState.
+--   If the file does not  exist at this path the benchmark will run, but be meaningless.
+--   See the README.md file in ledger-state/libs/ledger-state/README.md for instructions for
+--   how one might obtain such a file. Get your own copy and adjust the path here.
+newepochFilePath :: String
+newepochFilePath = "/home/sheard/Two/Data/71452796-NewEpochState.cbor"
+
+-- ==============================
+-- Stored constants read from files. Globals and NewEpochState
+
+instance NFData Globals where
+  rnf (Globals {}) = ()
+
+instance NFData (Filtered era) where
+  rnf (Filtered a b c d) = seq (rnf a) (seq (rnf b) (seq (rnf c) (rnf d)))
+
+readGlobals :: IO Globals
+readGlobals = pure globals
+  where
+    pp = def
+    !globals = mkGlobals shelleyGenesis pp
+
+readNewEpochState :: IO (NewEpochState CurrentEra)
+readNewEpochState = do
+  result <- try (Lazy.readFile newepochFilePath)
+  case result of
+    Right lazybytes ->
+      case decodeFullDecoder @(NewEpochState CurrentEra)
+        ("NewEpochState from " <> pack newepochFilePath)
+        fromCBOR
+        lazybytes of
+        Left err -> error (show err)
+        Right nes -> pure nes
+    Left e -> do
+      putStrLn $ "The path to the NewEpochState CBOR file\n   '" ++ newepochFilePath ++ "' is not readable."
+      putStrLn $ "Description of error\n   " ++ show (ioe_description e)
+      putStrLn $ "The error no.\n  " ++ show (ioe_errno e)
+      putStrLn $ "The benchmark results are meaningless."
+      pure bogusNewEpochState
+
+bogusNewEpochState :: NewEpochState CurrentEra
+bogusNewEpochState =
+  NewEpochState
+    (EpochNo 0)
+    (BlocksMade Map.empty)
+    (BlocksMade Map.empty)
+    def
+    (SJust (Complete (RewardUpdate (DeltaCoin 0) (DeltaCoin 0) def (DeltaCoin 0) def)))
+    (PoolDistr Map.empty)
+    def
+
+mkGlobals :: ShelleyGenesis CurrentEra -> Core.PParams CurrentEra -> Globals
+mkGlobals genesis pp =
+  mkShelleyGlobals genesis epochInfoE majorPParamsVer
+  where
+    majorPParamsVer = pvMajor $ _protocolVersion pp
+    epochInfoE =
+      fixedEpochInfo
+        (sgEpochLength genesis)
+        (mkSlotLength $ sgSlotLength genesis)
+
+-- =========================================================
+-- Functions based on Rules we can Benchmark
+
+updateRewardsX :: Globals -> NewEpochState CurrentEra -> EpochState CurrentEra
+updateRewardsX globals newepochstate =
+  let epochNo = nesEL newepochstate
+      epochstate = nesEs newepochstate
+   in case flip
+        runReader
+        globals
+        ( runRule
```

Comment:
This pattern is used twice. Also warrants a helper function

Reply:
Not sure what pattern we are taking about.

Reply:
I've learned it long ago. If I am doing a copy and paste of some chunk of code, especially in the same module, I will spend an extra moment to figure out how can I abstract away the common part in order to make it more readable and reduce future duplication

Reply:
done